src/include/internal/utility/alloc_impl.h

```diff
@@ -0,0 +1,194 @@
+/**
+ * @file
+ *
+ * @brief Helper for memory management.
+ *
+ * This file contains the implementations for the functions declared in
+ * <elektra/utility/alloc.h> and <internal/utility/alloc.h>
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#ifndef ELEKTRA_UTILITY_ALLOC_IMPL_H
+#define ELEKTRA_UTILITY_ALLOC_IMPL_H
+
+#include <internal/utility/assert.h>
+
+#include <limits.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef __cplusplus
+namespace ckdb
+{
+extern "C" {
+#endif
+
+/**Reallocate Storage in a save way.
+ *
+ *@code
+if (elektraRealloc ((void **) & buffer, new_length) < 0) {
+	// here comes the failure handler
+	// you can still use the old buffer
+#if DEBUG
+	fprintf (stderr, "Reallocation error\n");
+#endif
+	elektraFree (buffer);
+	buffer = 0;
+	// return with error
+}
+ *@endcode
+ *
+ * @param buffer is a pointer to a elektraMalloc
+ * @param size is the new size for the memory
+ * @retval -1 on failure
+ * @retval 0 on success
+ * @ingroup internal
+ */
+int elektraRealloc (void ** buffer, size_t size)
+{
+	ELEKTRA_ASSERT (size, "Size to allocate is zero (implementation defined behavior)");
+	void * ptr;
+	void * svr = *buffer;
+
+	ptr = realloc (*buffer, size);
+	ELEKTRA_ASSERT (ptr, "Memory (re)allocation failed with size %zu", size);
+	if (ptr == NULL)
+	{
+		// TODO: is this safe? couldn't the old buffer already be destroyed?
+		*buffer = svr; /* restore old buffer*/
+		return -1;
+	}
+	else
+	{
+		*buffer = ptr;
+		return 0;
+	}
```

Comment:
```suggestion
	ELEKTRA_ASSERT (size, "Size to allocate is zero (implementation defined behavior)");

	void * ptr = realloc (*buffer, size);
	ELEKTRA_ASSERT (ptr, "Memory (re)allocation failed with size %zu", size);
	if (ptr == NULL)
	{
		return -1;
	}

	*buffer = ptr;
	return 0;
```

Fix for https://github.com/ElektraInitiative/libelektra/pull/4941#discussion_r1215376240